src/gateway/ws.rs

```diff
@@ -75,25 +79,169 @@ struct WebSocketMessage<'a> {
+            } => {
+                const ZLIB_SUFFIX: [u8; 4] = [0x00, 0x00, 0xFF, 0xFF];
+
+                compressed.extend_from_slice(slice);
+                let length = compressed.len();
+
+                if length < 4 || compressed[length - 4..] != ZLIB_SUFFIX {
+                    return Ok(None);
+                }
+
+                let pre_out = inflater.total_out();
+                inflater
+                    .decompress(compressed, decompressed, flate2::FlushDecompress::Sync)
+                    .map_err(GatewayError::DecompressZlib)?;
+                compressed.clear();
+                let produced = (inflater.total_out() - pre_out) as usize;
+
+                Ok(Some(&decompressed[..produced]))
+            },
+
+            #[cfg(feature = "transport_compression_zstd")]
+            Compression::Zstd {
+                inflater,
+                decompressed,
+            } => {
+                let mut in_buffer = InBuffer::around(slice);
+                let mut out_buffer = OutBuffer::around(decompressed.as_mut());
+
+                let length = slice.len();
+                let mut processed = 0;
+
+                loop {
+                    match inflater.decompress_stream(&mut out_buffer, &mut in_buffer) {
+                        Ok(0) => break,
+                        Ok(_hint) => {},
+                        Err(code) => {
+                            return Err(Error::Gateway(GatewayError::DecompressZstd(code)))
+                        },
+                    }
+
+                    let in_pos = in_buffer.pos();
+
+                    let progressed = in_pos > processed;
+                    let read_all_input = in_pos == length;
+
+                    if !progressed {
+                        if read_all_input {
+                            break;
+                        }
+
+                        return Err(Error::Gateway(GatewayError::DecompressZstdCorrupted));
+                    }
+
+                    processed = in_pos;
+                }
+
+                let produced = out_buffer.pos();
+                Ok(Some(&decompressed[..produced]))
+            },
+        }
+    }
+}
+
+impl From<TransportCompression> for Compression {
+    fn from(value: TransportCompression) -> Self {
+        match value {
+            TransportCompression::None => Compression::Payload {
+                decompressed: Vec::new(),
+            },
+
+            #[cfg(feature = "transport_compression_zlib")]
+            TransportCompression::Zlib => Compression::Zlib {
+                inflater: ZlibInflater::new(true),
+                compressed: Vec::new(),
+                decompressed: vec![0; Self::DECOMPRESSED_CAPACITY].into_boxed_slice(),
+            },
+
+            #[cfg(feature = "transport_compression_zstd")]
+            TransportCompression::Zstd => {
+                let mut inflater = ZstdInflater::create();
+                inflater.init().expect("Failed to initialize Zstd decompressor");
```

Comment:
This kind of `create` and `init` pattern is making me iffy about `zstd_safe`, especially not being able to find any documentation on what it possibly errors on. Is there another library for it?

Reply:
<sub>this is restating things that have been said on discord for completeness sake.</sub>

[`zstd_safe`](https://github.com/gyscos/zstd-rs) is using bindings to the zstd reference implementation in C (https://github.com/facebook/zstd). The (to my knowledge) only other crate for zstd decompression that is not relying on `zstd_safe` in some way is [`ruzstd`](https://github.com/KillingSpark/zstd-rs). Looking at the docs for `ruzstd` it does not look that nice to work with in our context at least but maybe I'm wrong about that. The performance hit compared to  the C implementation is unclear, should probably be tested.